rfcs/chromium-connector.md

```diff
@@ -0,0 +1,213 @@
+
+#### User authentication in the CLI
+
+The authentication steps should be handled by the CLI using the actions ergonomics provided by the Node.js API.
+
+A typical login experience has a user and password `input`s and a submit `button` (everything might or not be inside a `form`). The user fills the inputs, pressed the button, and then the website navigates to another page or updates the URL with a hash. Puppeteer considers a navigation any change in the url, even if it is just a hash so this should cover the most common scenarios.
+
+The information required to do this programatically is:
+
+* **user value** and **`input` selector**
+* **password value** and **`input` selector**
+* **button selector** so it can be `click`ed
+
+In total 5 pieces of information.
+
+The most common ways to provide variables to a CLI are (from lower to higher priority):
+
+1. Environment variables
+1. Configuration file (`.hintrc`)
+1. CLI parameters
+1. A mix of the above: the user provides the selectors via `.hintrc` but uses environment variables for the credentials
+
+With this in mind, the options would be:
+
+| Entry Point | Values  |
+| ----------- | ------- |
+| ENV         | `WEBHINT_USER` <br> `WEBHINT_PASSWORD` <br> `WEBHINT_USER_SELECTOR` <br> `WEBHINT_PASSWORD_SELECTOR` <br> `WEBHINT_SUBMIT` |
+| CLI         | `user`, `-u` <br> `password`, `-p` <br> `user-selector`, `-us` <br> `password-selector`, `-ps` <br> `submit`, `-s`
+| Configuration | `user`, `password` and `submit` properties <br> in `options` with `selector` and `value`. <br> See below for example |
+
+`.hintrc` example:
+
+```json
+{
+  "connector": {
+    "name": "chromium",
+    "options": {
+       "user": {
+         "selector": "",
+         "value": ""
+       },
+       "password": {
+         "selector": "",
+         "value": ""
+       },
+       "submit": {
+         "selector": ""
+       }
+    }
+  }
+  // ...
+}
+```
+
+An example of a combination of these methods is:
+
+```json
+{
+  "connector": {
+    "name": "chromium",
+    "options": {
+       "user": {
+         "selector": "#user"
+       },
+       "password": {
+         "selector": "#pass"
+       },
+       "submit": {
+         "selector": "#login"
+       }
+    }
+  }
+  // ...
+}
+```
+```bash
+hint -u userName https://example.com
+```
+
+And then have the environment variable `WEBHINT_PASSWORD` set.
+
+**Limitations**
+
+The new API allows a user to analyze multiple websites simulatenously. This feature will be exposed also in the CLI. With this approach there are two problems:
+
+1. There is no way to provide different credentials to each URL
+1. There is no way to perform the authentication for only certain URLs
+
+These are edge cases and people in needs of this should use the Node.js API or re-architect their scripts to call `hint` multiple times with the right parameters in each case.
```

Comment:
Let's call out this should throw an error if someone tries to use the API with both authentication and multiple targets (unless there's some other handling we think is appropriate).

Reply:
If the user wants to analyze a bunch of pages in the same site behind authentication I think the CLI should support it. In fact, the Node.js API doesn't have any awareness of what `actions` is going to do so I don't see how we will be throwing an error in this case :/

Reply:
Ok.